### PR TITLE
Handling evars in the VM

### LINKED ISSUE #5659: vm_compute should handle evars
Note: the issue was created automatically with bugzilla2github tool

Original bug ID: BZ#5659
From: @JasonGross
Reported version: 8.6
CC:   @gmalecha, @ppedrot

### PR DESCRIPTION
**Pull request by ppedrot**
This is actually pretty straightforward. We simply treat them as as an application of an atom to its instance,
and in the decompilation phase we reconstruct the instance from the arguments.

This closes #5659.

Note that we could probably do the same in native compilation for the sake of uniformity, but this is left to another PR.

### CODE REVIEW

**Comment by maximedenes** [comment]
@ppedrot I believe it is already done in `native_compute` :)

**Comment by ppedrot** [comment]
@maximedenes *Au temps pour moi* as we say.

**Comment by ppedrot** [comment]
Also fixes [BZ#5060](https://coq.inria.fr/bugs/show_bug.cgi?id=5060).

**Comment by silene** [comment]
I have said it several times already, but in case someone missed it, here it goes again. I think it is a very bad idea to support evars that way. One of the desirable features of `vm_compute` is that its behavior is completely predictable both at tactic time and at `Qed` time (and for `vm_cast_no_check`, at `Qed` time). Adding support for evar in the virtual machine means that `vm_compute` becomes unpredictable at `Qed` time. We will soon get bug report saying that `vm_compute` is instantaneous at tactic time but `Qed` never terminates. Ideally, evars (and opaque terms and anything that should not be reduced, e.g. the goal for `vm_compute in`) should be abstracted away from a term before the virtual machine is called.

**Comment by ppedrot** [comment]
@silene That's true, but I think that's something that should be done at tactic level, not at kernel level.

**Comment by silene** [comment]
> That's true, but I think that's something that should be done at tactic level, not at kernel level.

That is kind of my point. You are adding a lot of code to the kernel yet breaking a feature of `vm_compute`. This whole code would be useless if the issue with `vm_compute` was fixed properly.

**Comment by ppedrot** [comment]
@silene A lot of code is maybe exaggerated, that actually quite small. That said, I do hear your complaint, and I can submit an alternative patch if you want. Note that we ought to check that all of the entry-points to the VM are indeed evar-free, as I'm not sure that vm_compute is the sole way to call the VM.

**Comment by ppedrot** [comment]
@silene So I've thought a bit about it, and now I think that this is the proper way of handling evars in the VM anyways.

Indeed, you could mangle the term in the VM normalization to extract out the evars, but it's not going to help anyway in your case, because the cast is not inserted by the normalization function in `pretyping/vnorm`. It's instead generated by the conversion tactic from `tactics/tactics`. And therefore this should be that part of the code that mangles the term, in order to recover the Qed-stability property you long for.

But then you run into trouble, because you must ensure that all callers of the VM normalization function are going to do this kind of mangling, although it doesn't always make sense. For instance, merely computing the normalization through `eval vm_compute in` has no place to put a cast in sight.

Therefore, I think this patch should go as-is, but that we should also provide another patch that specifically targets the `vm_compute` tactic (and *not* the normalization function) to be slightly more clever than the normalization function.

**Comment by silene** [comment]
I understand your point about `eval vm_compute in` and I agree with this part of the patch. I still disagree about the removal of the error message when using `vm_compute` on a goal with evars. We are shooting our own foot since it will almost never behave the way the user would expect at `Qed` time.

**Comment by ppedrot** [comment]
It's already the case that `vm_compute in` is broken, and nonetheless it is still useful, disregarding [BZ#4637](https://coq.inria.fr/bugs/show_bug.cgi?id=4637). But I'll try to come up with a patch, don't worry.

**Comment by JasonGross** [comment]
> We are shooting our own foot since it will almost never behave the way the user would expect at Qed time.

The way I expect it to behave is that if I have an evar that is one of the arguments to a constructor or to a type family, the performance should not change between the proof script and the `Qed`. In particular, the only use case I have is for a goal `x = ?e` or `?e = x`, for `x` evar-free, I want to be able to do `vm_compute; reflexivity`.  Perhaps add an `evm_compute` variant that permits evars, and have `vm_compute` be evar-free?

**Comment by silene** [comment]
> I have is for a goal x = ?e or ?e = x, for x evar-free, I want to be able to do vm_compute; reflexivity

If you do that, `Qed` will be twice as slow. That is what I mean when I say that `vm_compute` on evars does not behave the way you expect. The proper way is something which looks as follows:

    match goal with
    | |- _ = ?m =>
      let p := eval vm_compute in m in
      instantiate (1 := p) ;
      vm_cast_no_check (eq_refl p)
    end.

**Comment by JasonGross** [comment]
> If you do that, `Qed` will be twice as slow

Why will `Qed` be twice as slow?  (Also, is there a difference between `instantiate (1:=p)` and `unify ?e p` and `unify p ?e`, speed-wise?)

In the example you gave, the proof term should be `eq_refl p <: p = m`, right?

If you do `vm_compute; reflexivity`, given `?Goal : ?e = m`, the `vm_compute` changes the proof term to `?Goal1 <: ?e = m` of type `?e = m` where `?Goal1 : ?e = p`, and then `reflexivity` instantiates `?e` with `p` and `?Goal1` with `eq_refl p`, changing the overall proof term to `eq_refl p <: p = m`, which is the same as the code you gave.  The cost of doing `vm_compute` in my version should be the same as the cost of doing `eval vm_compute in` + `vm_cast_no_check` in your version, and the cost of doing `reflexivity` in my version should be the same as the cost of doing `instantiate (1:=p)` + typing the constr `eq_refl p` in your version.  Did I mess up my analysis?

**Comment by silene** [comment]
> `eq_refl p <: p = m`

My mistake. For some reason, I thought that the resulting proof term would be `eq_refl p <: m = m` (hence the doubled time). So you are right, this is a case where `vm_compute` behaves properly with evars and both methods will have exactly the same complexity.

**Comment by ppedrot** [comment]
@silene Do you agree that the slow behaviour at Qed-time is a bug, but which is independent from the VM, and this PR in particular? All reduction functions display the same issue, not only vm_compute, and the fact the latter was erroring on terms containing evars was a mere lucky fact. If so, we could first merge this PR and then come up with a general patch for reduction functions, which is not as easy as it sounds.

**Comment by ppedrot** [comment]
@silene Ping.

**Comment by ppedrot** [comment]
@silene Ping again.

**Comment by silene** [comment]
No I don't agree that all reduction functions display the same issue. Only `vm_compute` and `native_compute` insert a specific annotation that forces the kernel to perform normalization, which is not stable by evar substitution. The other reduction functions do not ask for normalization, but only conversion. I much prefer the suggestion of @JasonGross, that is to say, force the users to explicit tell they know what they are doing.

**Comment by bgregoir** [comment]
I have read most of the comment (quickly) and the code
The original intention of not having the evar was that evar are not part of the system (I mean
what should be the kernel). 
I was completely agree with Guillaume's argument on 
"One of the desirable features of vm_compute is that its behavior is completely predictable both at tactic time and at Qed time"
Then I realised that the feature is already broken, since if you do a proof with an "assert (x: )"
then x, will be opaque during the tactic and will be transparent during the Qed.
It can be a good idea to also patch the let to make this distinction explicite.
I also think that one day we will have a request for the Qed time 
(but we are already waiting for the one with the assert)

I have read the code, and it seems reasonable, not to big and keeping the spirit.

So in conclusion, I think it can be integrated (but I am not a current developper)

**Comment by silene** [comment]
> Then I realised that the feature is already broken, since if you do a proof with an "assert (x: )" then x, will be opaque during the tactic and will be transparent during the Qed.

Do you have an example? I tried a bit and I never got `assert` to interfere with `vm_compute`, since the opaque content occurs in values and not in types, and thus does not reach the VM. Example:

```coq
Goal True.
Proof.
vm_compute.
assert (n := fact 30).
now cut (n = n).
Qed.
```

**Comment by bgregoir** [comment]
Yes. 
The point is that we have to to something that need a little be of computation.
i.e. When the vm is called to check convertibility nothing is done if the conversion is trivial (equality of the to term).
For example if we do
Goal True.
assert (n := fact 30).
cut (n = n).
+ trivial.
vm_compute.
reflexivity.
Qed.

The proof term is 
let n := fact 30 in
let H := (eq_refl <: n = n) : n = n in (fun _ : n = n => I) H
so the vm is used to check that the type of (@eq_refl _ n) i.e n = n is convertible with n = n.
Since the convertibility check is trivial the vm is no used.

Now we can do the following version.
Goal True.
assert (n := fact 30).
cut (1 + n = S n).
+ trivial.
vm_compute.
reflexivity.
(* Upto that point every thing is immediate *)
Qed.
(* I let you wait for the end *)
The point is that during the proof mode the vm is used to check that
S n = S n is convertible with 1 + n = S n.
And is it trivial to do when n is a variable.
But during "Qed", n is "fact 30", so the normal S n is something like
2.6525286e+32 + 1 and we have to compute it 4 times (this does not really change the problem but ...)

**Comment by JasonGross** [comment]
@ppedrot Does this still need discussing?  My understanding of the consensus is that if we get both `vm_compute ` and `evm_compute ` then @silene will be happy with this PR.

**Comment by ppedrot** [comment]
I can indeed add a new tactic `evm_compute`, if that's what people think is best.

I'm not really sure it is though, as it defines yet another built-in tactic (plus specific stuff to make `eval ... in` work), and furthermore `native_compute` doesn't complain about evars albeit it suffers from the same issue. Furthermore it makes this new tactic not quite atomic, as checking for evars can be done with another built-in tactic.

I'll rebase and push a commit atop of this branch with the `evm` variant.

**Comment by maximedenes** [comment]
I would rather make `vm_compute` accept evars because 1- it will be uniform with `native_compute` (I plan to unify the normalization and conversion code of both) and 2- accepting evars seems easier for beginners, a tactic checking the absence of evars can be used when more control is needed.

**Comment by maximedenes** [comment]
WG discussion: @ppedrot & @maximedenes think `vm_compute` should handle evars.

File: pretyping/vnorm.ml [hunk 1]
```diff
@@ -352,7 +373,7 @@ and nf_cofix env sigma cf =
   mkCoFix (init,(name,cft,cfb))
 
 let cbv_vm env sigma c t  =
-  if Termops.occur_meta_or_existential sigma c then
+  if Termops.occur_meta sigma c then
     CErrors.user_err Pp.(str "vm_compute does not support existential variables.");
```
Comment (maximedenes) [inline pretyping/vnorm.ml]: Please change "existential variables" to "metavariables" in the message.

**Review by maximedenes — DISMISSED** [review]

**Comment by maximedenes** [comment]
So @silene, is it ok if we merge this patch that adds support for evars to the vm reduction machinery, and then discuss separately whether the `vm_compute` tactic should be restricted?

**Comment by silene** [comment]
Sure. Since no one else thinks that existential variables inside bytecode/native execution are a footgun, I should not be blocking this pull request.

File: pretyping/vnorm.ml (hunk 1 above)
Comment (herbelin) [reply pretyping/vnorm.ml]: ...this reminds me that the reference manual would probably be clearer if it could make up its mind between the terminology metavariable vs existential variables (or at least once for all say somewhere that they are essentially synonym).

Comment (herbelin) [reply pretyping/vnorm.ml]: Hum, I spoke too quickly. The reference manual is actually ok (except in the ssr chapter which uses the cute terminology of "existential metavariables"). Otherwise, the terminology is stable:
- metavariables = Ltac pattern-matching variables
- existential variables = holes in terms
